File: usaco-abd/gift1.go
package main

import (
	"os"
	"fmt"
)

/*

compared to the solution, two things is bad:
no oo at all, and no symbol abstraction of people name at all
and you should parpare for string reuse things


*/


func main() {
	temp := make([]byte, 0)

	fin, _ := os.Open("gift1.in")
	defer fin.Close()
	fout, _ := os.Create("gift1.out")
	defer fout.Close()

	n := 0
	fmt.Fscanf(fin, "%d", &n)
	friends := make(map[string]int)
	names := make([]string, n, n)

	for i := 0; i < n; i++ {
		fmt.Fscan(fin, &temp)
		names[i] = string(temp)
		friends[names[i]] = 0
	}

	for i := 0; i < n; i++ {
		money, part := 0, 0
		fmt.Fscan(fin, &temp, &money, &part)
		name := string(temp)
		if part == 0 {
			continue
		}
		friends[name] = friends[name] - money + money % part
		for j := 0 ; j < part; j++ {
			fmt.Fscan(fin, &temp)
			name := string(temp)
			friends[name] = friends[name] + money / part
		}
	}
	for _, val := range names {
		fout.WriteString(fmt.Sprintf("%s %d\n", val, friends[val]))
	}
}

File: USACO/gift1.cpp
/*
ID: pirripe1
PROG: gift1
LANG: C++
*/
#include <iostream>
#include <fstream>
#include <string>
#include <cstdlib>
#include <map>
#include <queue>

using namespace std;

int main() {
    ofstream fout("gift1.out");
    ifstream fin("gift1.in");

    int n;
    fin >> n;

    map<string, int> friends;

    queue<string> names;
    for (int i = 0; i < n; i++) {
        string str;
        fin >> str;
        names.push(str);
        friends[str] = 0;
    }


    for (int i = 0; i < n; i++) {
        string name;
        int money, part;
        fin >> name >> money >> part;
        if (part == 0) {
            continue;
        }
        friends[name] = friends[name] - money + money % part;
        for (int j = 0; j < part; j++) {
            string name;
            fin >> name;
            friends[name] = friends[name] + money / part;
        }
    }
    for (int i = 0; i < n; i++) {
        string name = names.front();
        names.pop();
        fout << name << ' ' << friends[name] << endl;
    }

    fin.close();
    fout.close();
    return 0;
}